File: Princy_Ballabh/Day_20.java
public class Day_20{
    public static void main(String args[]){
        int n=4;
        int arr[]={10,20,30,40};
        int k=2;
        System.out.println("The minimum time to paint all boards: "+mintime(arr,n,k));
    }

    private static int sum(int arr[], int s, int e){
        int tot=0,i;
        for(i=s;i<=e;i++)
            tot+=arr[i];
        return tot;
    }

    public static int mintime(int arr[], int n, int k){
        int dp[][]=new int[k+1][n+1];
        int i,j,p;
        for(i=1;i<=n;i++){
            dp[1][i]=sum(arr,0,i-1);
        }
        for(i=1;i<=k;i++){
            dp[i][1]=arr[0];
        }
        for(i=2;i<=k;i++){
            for(j=2;j<=n;j++){ 
                int best=Integer.MAX_VALUE;
                for(p=1;p<=j;p++){
                    best=Math.min(best,Math.max(dp[i-1][p],sum(arr,p,j-1)));
                }
                dp[i][j]=best;
            }
        }
        return dp[k][n];
    }  
}

File: Mohammad-Abbas/Day20.cpp
 #include <iostream>
#include <vector>
#include <climits>

using namespace std;

// Function to calculate the sum of elements in arr from start to end
int sum(vector<int>& arr, int start, int end) {
    int total = 0;
    for (int i = start; i <= end; i++) {
        total += arr[i];
    }
    return total;
}

int minTimeToPaintBoards(vector<int>& arr, int n, int k) {
    // DP table
    vector<vector<int>> dp(n + 1, vector<int>(k + 1, INT_MAX));

    // Base case: with one painter, the time is the sum of the first i boards
    for (int i = 1; i <= n; i++) {
        dp[i][1] = sum(arr, 0, i - 1);
    }

    // Fill the DP table
    for (int j = 2; j <= k; j++) { // For each painter from 2 to k
        for (int i = 1; i <= n; i++) { // For each board from 1 to n
            for (int p = 1; p < i; p++) { // Try to partition the boards
                dp[i][j] = min(dp[i][j], max(dp[p][j - 1], sum(arr, p, i - 1)));
            }
        }
    }

    return dp[n][k];
}

int main() {
    int n = 4;
    vector<int> arr = {10, 20, 30, 40};
    int k = 2;

    int result = minTimeToPaintBoards(arr, n, k);
    cout << "Minimum time to paint all boards: " << result << endl;

    return 0;
}
